### ISSUE TITLE
How To Config with DllReferencePlugin

### ISSUE THREAD

**Issue by GuillaumeCisco** [SimenB/add-asset-html-webpack-plugin#7]
Hello, thank you for this plugin.
I'm sorry to open issue for a question, but I don't know where to ask this.

I've tried to set up the `AddAssetHtmlPlugin` with the `DllReferencePlugin` but I don'think it works correctly.
For information I have a folder `webpack` with ebpacks configs and a folder `dll` which will contain dll, manifest and sources files. Thes two folders are in the same parent folder.

My webpack config for generating dll, manifest and source maps looks like this :
```js
import webpack from 'webpack';

export default {
    entry: {
        Babel: [
            "babel-cli",
            ...
        ],
        React: [
            "react",
            ...
        ]
    },
    devtool: '#source-map',
    output: {
        path: `${__dirname}/../dll`,
        filename: '[name].dll.js',
        library: '[name]_[hash]',
    },

    plugins: [
        new webpack.DllPlugin({
            name: '[name]_[hash]',
            path: `${__dirname}/../dll/[name]-manifest.json`,
            context: `${__dirname}/../dll`
        })
    ]
};
```

And my webpack configuration is :
```js
import ExtractTextPlugin from 'extract-text-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import AddAssetHtmlPlugin from 'add-asset-html-webpack-plugin';
import loaders from './loaders';
import resolve from './resolve';
import {definePlugin} from './plugins';
import webpack from 'webpack';
var DEBUG = !(['production', 'development'].includes(process.env.NODE_ENV));
var DEVELOPMENT = process.env.NODE_ENV === 'development';
var PRODUCTION = process.env.NODE_ENV === 'production';
var precss = require('precss');
var autoprefixer = require('autoprefixer');
var cssnano = require('cssnano');


export default {
    entry: {
        index: [
            `${__dirname}/../src/frontend/js/main.js`,
            `${__dirname}/../src/frontend/css/main.scss`,
        ]
    },
    module: {
        loaders: loaders('frontend'),
    },
    postcss: function () {
        return [autoprefixer({browsers: ['last 2 versions']}), precss, cssnano];
    },
    stats: {
        colors: true,
        reasons: DEBUG,
        hash: DEVELOPMENT,
        version: DEVELOPMENT,
        timings: true,
        chunks: DEVELOPMENT,
        chunkModules: DEVELOPMENT,
        cached: DEVELOPMENT,
        cachedAssets: DEVELOPMENT,
    },
    output: {
        filename: '[name].js',
        path: `${__dirname}/../build/frontend`,
        publicPath: '/'
    },
    devtool: DEBUG ? 'source-map' : (DEVELOPMENT ? '#cheap-module-eval-source-map' : '#hidden-source-map'),
    plugins: [
        definePlugin(),
        new webpack.DllReferencePlugin({
            context:  `${__dirname}/..`,
            manifest: require(`${__dirname}/../dll/Babel-manifest.json`)
        }),
        new webpack.DllReferencePlugin({
            context: `${__dirname}/..`,
            manifest: require(`${__dirname}/../dll/React-manifest.json`)
        }),
        new HtmlWebpackPlugin({
            filename: 'index.html',
            favicon: `${__dirname}/../src/frontend/favicon.ico`,
            template: `${__dirname}/../src/frontend/index.html`,
            hash: true
        }),
        new AddAssetHtmlPlugin({
            filename: require.resolve(`${__dirname}/../dll/Babel.dll.js`),
            includeSourcemap: true
        }),
        new AddAssetHtmlPlugin({
            filename: require.resolve(`${__dirname}/../dll/React.dll.js`),
            includeSourcemap: true
        }),        
    ],
    resolve: resolve(),
    devServer: {
        historyApiFallback: true,
    },
    watch: true
};
```

When I run my webpack dev server with a config including HMR, and going to a page, It looks like nothing changed at all... Build time is the same... nearly 15-20 seconds.
And I find nothing in the index.html generated regarding manifest files.

Only things I can see are:
```
webpack: wait until bundle finished: /headband/
Bundled in 20611ms!
Hash: 51d529ec8ec7a907d22b
Version: webpack 1.12.14
Time: 20613ms
      Asset       Size  Chunks             Chunk Names
   index.js    13.6 MB       0  [emitted]  index
favicon.ico  515 bytes          [emitted]  
 index.html  467 bytes          [emitted]  
chunk    {0} index.js (index) 5.01 MB [rendered]
    [0] multi index 64 bytes {0} [built]
    [1] ./~/react/react.js 56 bytes {0} [built]
    [2] ./~/babel-polyfill/~/core-js/modules/_export.js 1.6 kB {0} [built]
    [3] ./~/babel-runtime/helpers/extends.js 555 bytes {0} [built]
    [4] ./~/moment/moment.js 119 kB {0} [built]
    ......
    [1643] ./~/url/util.js 314 bytes {0} [built]
    [1644] (webpack)-dev-server/client?http://localhost:3000/ 2.67 kB {0} [built]
    [1645] (webpack)/buildin/amd-options.js 43 bytes {0} [built]
    [1646] (webpack)/hot/dev-server.js 1.85 kB {0} [built]
    [1647] (webpack)/hot/log-apply-result.js 813 bytes {0} [built] 
   Child html-webpack-plugin for "index.html":
         Asset     Size  Chunks       Chunk Names
    index.html  21.5 kB       0       
    chunk    {0} index.html 393 bytes [rendered]
        [0] ./~/html-webpack-plugin/lib/loader.js!./src/frontend/index.html 393 bytes {0} [built]
webpack: bundle is now VALID.
```

For more informations, my loaders file looks like (note the `exclude: /node_modules/`):
```js
import ExtractTextPlugin from 'extract-text-webpack-plugin';
var DEBUG = !(['production', 'development'].includes(process.env.NODE_ENV));
var DEVELOPMENT = process.env.NODE_ENV === 'development';
var PRODUCTION = process.env.NODE_ENV === 'production';

export default function (appName) {
    const prefix = appName ? `${appName}/` : '';

    const loaders = [{
        test: /\.jsx?$/,
        exclude: /node_modules/,
        loader: 'babel-loader',
        // Options to configure babel with
        query: {
            plugins: [
                'transform-runtime',
                ...(PRODUCTION ? [
                    "transform-react-constant-elements",
                    "transform-react-inline-elements",
                    "transform-react-remove-prop-types"
                ] : [])
            ],
            presets: [
                "react",
                "es2015",
                "stage-0",
            ]
        },
    }, {
        test: /\.json$/,
        loader: 'json-loader',
    }, {
        test: /\.jpe?g$|\.gif$|\.png$/,
        loader: `url-loader?limit=10000&name=/${prefix}[hash].[ext]`,
    }, {
        test: /\.(otf|svg)(\?.+)?$/,
        loader: 'url-loader?limit=8192',
    }, {
        test: /\.eot(\?\S*)?$/,
        loader: 'url-loader?limit=100000&mimetype=application/vnd.ms-fontobject',
    }, {
        test: /\.woff2(\?\S*)?$/,
        loader: 'url-loader?limit=100000&mimetype=application/font-woff2',
    }, {
        test: /\.woff(\?\S*)?$/,
        loader: 'url-loader?limit=100000&mimetype=application/font-woff',
    }, {
        test: /\.ttf(\?\S*)?$/,
        loader: 'url-loader?limit=100000&mimetype=application/font-ttf',
    }, {
        test: /\.html$/,
        loader: 'html',
    }, {
        test: /\.s?css$/,
        ...(PRODUCTION ? {loader: ExtractTextPlugin.extract('style-loader', ['css-loader', 'postcss-loader'])} : {loaders: ["style", "css", "sass"]} )
    }];

    return loaders;
}
```

And my resolve file :
```js
import path from 'path';

export default function () {
    return {
        root: path.resolve(`${__dirname}/..`),
        alias: {
            isomorphic: 'src/isomorphic',
        },
    };
}
```
I don't understand what I miss.
Some help will be very welcome here.
Thank you very much,

**Comment by GuillaumeCisco** [comment]
Update:
Found that replacing `html-webpack-plugin-before-html-generation` by `html-webpack-plugin-before-html-processing` write correctly the assets in my index.html file.
But I see no improvements in webpack-server-dev start, nor HMR, no building...
I feel fooled.

**Comment by GuillaumeCisco** [comment]
Update 2:
Build seems really faster as I removed `new webpack.optimize.AggressiveMergingPlugin(),` which apprently is incompatible with DLL Plugin.

**Comment by SimenB** [comment]
If the app still worked even though thw script wasn't included in the html, it means that the DllReferencePlugin isn't doing its job. 

Got a repo I could take a look at?

**Comment by GuillaumeCisco** [comment]
Hello @SimenB ,

Thank you for caring.
I've created a light version of my project by removing all business code. That's why some business logic files will look very empty :)

You can find the repo here: https://github.com/GuillaumeCisco/dllPluginTest

As you will see, I import your library in my package.json, but don't use it, instead I use a custom file based from your library with the replacment of `html-webpack-plugin-before-html-generation` by `html-webpack-plugin-before-html-processing`.

Thank you very much :) :+1:

**Comment by SimenB** [comment]
@GuillaumeCisco haven't checked too closely, but you need a newer `html-webpack-plugin`, the event I use was added in `2.10.0` https://github.com/ampedandwired/html-webpack-plugin/blob/master/CHANGELOG.md#v2100

**Comment by SimenB** [comment]
@GuillaumeCisco Opened PR on your project

**Comment by SimenB** [comment]
@GuillaumeCisco did it fix your problem? If so, any ideas on how to improve the docs to make it easier to understand?

**Comment by SimenB** [comment]
Closed in GuillaumeCisco/dllPluginTest#1
